core/src/common.cpp

```diff
@@ -362,6 +364,439 @@ namespace irods::http
 		return std::nullopt;
 	}
 
+	/// Validates an OAuth 2.0 Access Token using the Introspection Endpoint
+	/// See RFC 7662 for more details
+	///
+	/// \returns An optional containing a nlohmann::json object if verification was successful. Otherwise,
+	///          an empty optional is returned.
+	auto validate_using_introspection_endpoint(std::string _bearer_token) -> std::optional<nlohmann::json>
+	{
+		namespace logging = irods::http::log;
+
+		body_arguments args{{"token", _bearer_token}, {"token_type_hint", "access_token"}};
+
+		auto json_res{hit_introspection_endpoint(url_encode_body(args))};
+
+		// Validate access token
+		if (!json_res.at("active").get<bool>()) {
+			logging::warn("{}: Access token is invalid or expired.", __func__);
+			return std::nullopt;
+		}
+
+		return json_res;
+	}
+
+	/// Fetches JWKs from the location specified by the OpenID Provider
+	///
+	/// See OpenID Connect Discovery 1.0 Section 3 for info on jwks_uri
+	/// See RFC 7517 for more information on JSON Web Key (JWK)
+	///
+	/// \returns A std::string represting the JWKs from the OpenID Provider
+	///
+	/// \todo Stash the results from this function, only allow to run once.
```

Comment:
Remove TODO, static acts as stash.

At most, consider moving to OIDC init. This will grow a new global, and it is unlikely the `jwks_uri` is improperly configured.

Additionally, having it as it is now could allow for retry behavior?

Reply:
Done.